File: $_algotsu-kaizen/problemo-set/lexicographical-numbers.c

void fill(int i, int n, int* list, int* idx) {
    if (i > n) return;
    *(list + (*idx)++) = i;
    fill(i * 10, n, list, idx);
    if (i % 10 < 9) fill(i + 1, n, list, idx);
}

int* lexicalOrder(int n, int* returnSize) {
    int* result = (int *)malloc(sizeof(int) * n);
    *returnSize = n;
    int idx = 0;
    fill(1, n, result, &idx);
    return result;
}

File: $_algotsu-kaizen/problemo-set/lexicographical-numbers.rs
impl Solution {

    fn generate(i: i32, n: i32) -> Vec<i32> {
        let mut result = Vec::new();
        if i > n { return result };
        result.push(i);
        result.append(&mut Self::generate(i * 10, n));

        if i % 10 < 9 {
            result.append(&mut Self::generate(i + 1, n));
        }

        return result;
    }

    pub fn lexical_order(n: i32) -> Vec<i32> {
        return Self::generate(0b1, n);
    }
}
